mmf/models/lxmert.py

```diff
@@ -0,0 +1,800 @@
+            pooled_output = pooled_output.view(-1, pooled_output.size(1) * 2)
+
+        logits = self.classifier(pooled_output)
+        reshaped_logits = logits.contiguous().view(-1, self.config.num_labels)
+        output["scores"] = reshaped_logits
+
+        return output
+
+
+@registry.register_model("lxmert")
+class LXMERT(BaseModel):
+    def __init__(self, config):
+        super().__init__(config)
+
+    @classmethod
+    def config_path(cls):
+        return "configs/models/lxmert/pretrain.yaml"
+
+    # Backward compatibility
+    @classmethod
+    def format_state_key(cls, key):
+        return (
+            key.replace("bert.bert", "model.bert")
+            .replace("bert.cls", "model.cls")
+            .replace("bert.classifier", "model.classifier")
+        )
+
+    def build(self):
+        if self.config.training_head_type == "pretraining":
+            self.model = LXMERTForPretraining(self.config)
+        else:
+            self.model = LXMERTForClassification(self.config)
+
+        if getattr(self.config, "freeze_base", False):
+            for p in self.model.bert.parameters():
+                p.requires_grad = False
+
+    def get_image_and_text_features(self, sample_list):
+        # bert input
+        bert_input_ids = sample_list.input_ids
+        bert_input_mask = sample_list.input_mask
+        bert_input_type_ids = sample_list.segment_ids
+        masked_lm_labels = sample_list.lm_label_ids
+        # image input
+        image_info = getattr(sample_list, "image_info_0", {})
+        image_dim_variable = getattr(image_info, "max_features", None)
+        image_feature_variable = getattr(sample_list, "image_feature_0", None)
+        if image_feature_variable is None:
+            raise Exception("object features must be provided for lxmert")
+        bbox = getattr(image_info, "bbox", None)
+        if bbox is None:
+            raise Exception("bounding boxes must be provided for lxmert")
+        else:
+            bbox = torch.tensor(bbox)
+            max_features = torch.tensor(
+                min(image_feature_variable.shape[1], bbox.shape[1]), dtype=torch.int
+            ).cuda()
+            bbox = bbox[:, : max_features.item(), :4]
+            image_h = getattr(image_info, "image_height", None)
+            image_w = getattr(image_info, "image_width", None)
+            if image_h is not None and image_w is not None:
+                image_h = torch.tensor(image_h, dtype=torch.int)
+                image_w = torch.tensor(image_w, dtype=torch.int)
+                image_location = torch.zeros(tuple(bbox.shape))
+                image_location[:, :, 0] /= image_w[:, None]
+                image_location[:, :, 1] /= image_h[:, None]
+                image_location[:, :, 2] /= image_w[:, None]
+                image_location[:, :, 3] /= image_h[:, None]
+            else:
+                image_location = bbox
+            image_location_variable = image_location.cuda()
```

Comment:
Use `.to(device)` with proper device from registry. This will not allow LXMERT to run on CPU. Please take care of this at other cuda calls as well.

Reply:
Alternatively, you can `torch.new_*` to create tensors with same dtype as input to avoid figuring out device manually.